pkg/controller/hibernation/hibernation_controller.go

```diff
@@ -148,6 +149,63 @@ func (r *hibernationReconciler) Reconcile(request reconcile.Request) (reconcile.
 		return reconcile.Result{}, nil
 	}
 
+	// Check for the hibernate-after annotation, and if the cluster has been in running state for longer than this duration, put it to sleep.
+	hibernateAfter, ok := cd.Annotations[constants.HibernateAfterAnnotation]
+	if ok && cd.Spec.PowerState != hivev1.HibernatingClusterPowerState {
+		cdLog.Debugf("found hibernate-after annotation: %s", hibernateAfter)
+		dur, err := time.ParseDuration(hibernateAfter)
+		if err != nil {
+			cdLog.WithError(err).WithField("hibernateAfter", hibernateAfter).Infof("error parsing %s as a duration", constants.HibernateAfterAnnotation)
+			return reconcile.Result{}, fmt.Errorf("error parsing %s as a duration: %v", constants.HibernateAfterAnnotation, err)
+		}
+
+		runningSince := cd.Status.InstalledTimestamp.Time
+
+		cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.ClusterHibernatingCondition)
+		if cond == nil {
+			log.WithField("runningSince", runningSince).Info("cluster has no hibernating condition (never hibernated), using installed time")
+		} else if cond.Status == corev1.ConditionFalse {
+			runningSince = cond.LastTransitionTime.Time
+			cdLog.WithField("runningSince", runningSince).WithField("reason", cond.Reason).Info("hibernating condition false")
+		} else {
+			// Cluster is Hibernating, we have no runningSince time:
+			runningSince = time.Time{}
+		}
+
+		hibLog := cdLog.WithFields(log.Fields{
+			"runningSince":   runningSince,
+			"hibernateAfter": hibernateAfter,
+		})
+
+		if !runningSince.IsZero() {
+			expiry := runningSince.Add(dur)
+			hibLog.Debugf("cluster should be hibernating after: %s", expiry)
+			if time.Now().After(expiry) {
+				hibLog.WithField("expiry", expiry).Info("cluster has been running longer than hibernate-after duration, moving to hibernating powerState")
+				cd.Spec.PowerState = hivev1.HibernatingClusterPowerState
+				err := r.Update(context.TODO(), cd)
+				if err != nil {
+					hibLog.WithError(err).Log(controllerutils.LogLevel(err), "error hibernating cluster")
+				}
+				return reconcile.Result{}, err
+			}
+
+			defer func() {
+				requeueNow := result.Requeue && result.RequeueAfter <= 0
+				if returnErr == nil && !requeueNow {
+					// We have an hibernate after time but cluster has not been running that long yet.
+					// Set requeueAfter for just after so that we requeue cluster for hibernation once reconcile has completed
+					requeueAfter := time.Until(expiry) + 60*time.Second
+					if requeueAfter < result.RequeueAfter || result.RequeueAfter <= 0 {
+						hibLog.Infof("cluster will reconcile due to hibernate-after time in: %v", requeueAfter)
+						result.RequeueAfter = requeueAfter
```

Comment:
Set `result.Requeue` to `true` also to cover the scenario (unlikely with the 60-second addition) that `requeueAfter` is not positive.